src/main/scala/chisel3/util/HeterogeneousVec.scala

```diff
@@ -0,0 +1,112 @@
+// See LICENSE for license details.
+
+package chisel3.util
+
+import chisel3._
+import chisel3.core.{Data, requireIsChiselType, requireIsHardware}
+import chisel3.internal.naming.chiselName
+
+import scala.collection.immutable.ListMap
+
+object HeterogeneousVecWireInit {
+  /**
+    * Construct a new wire with the given bound values.
+    * This is analogous to [[chisel3.core.VecInit]].
+    * @param vals Values to create a HeterogeneousVec with and assign
+    * @return HeterogeneousVec with given values assigned
+    */
+  def apply[T <: Data](vals: Seq[T]): HeterogeneousVec[T] = {
+    // Create a wire of this type.
+    val hetVecWire = Wire(HeterogeneousVec(vals.map(_.cloneTypeFull)))
+    // Assign the given vals to this new wire.
+    for ((a, b) <- hetVecWire.zip(vals)) {
+      a := b
+    }
+    hetVecWire
+  }
+}
+
+object HeterogeneousVec {
+  /**
+    * Create a HeterogeneousVec from the type of the given Vec.
+    * For example, given a Vec(2, UInt(8.W)), this creates HeterogeneousVec(Seq.fill(2){UInt(8.W)}).
+    * @param vec Vec to use as template
+    * @return HeterogeneousVec analogous to the given Vec.
+    */
+  def apply[T <: Data](vec: Vec[T]): HeterogeneousVec[T] = {
+    HeterogeneousVec(Seq.fill(vec.length)(vec.sample_element))
+  }
+}
+
+/**
+  * A hardware array of elements that can hold values of different types/widths,
+  * unlike Vec which can only hold elements of the same type/width.
+  *
+  * @param eltsIn Element types. Must be Chisel types.
+  *
+  * @example {{{
+  * val v = Wire(HeterogeneousVec(Seq(UInt(8.W), UInt(16.W), UInt(32.W))))
+  * v(0) := 100.U(8.W)
+  * v(1) := 10000.U(16.W)
+  * v(2) := 101.U(32.W)
+  * }}}
+  */
+@chiselName
+case class HeterogeneousVec[T <: Data](private val eltsIn: Seq[T]) extends Record with collection.IndexedSeq[T] {
```

Comment:
This should not be a `case class` but probably be final.

`case class` has lots of implications about equality and immutability that really don't fit with Chisel types

Reply:
Generally allowing a class to be extended has a cost if we want to make any changes in the future. I think start with it as `final` and then if people ask for it to be make extendible we can evaluate it at a later time.

Reply:
Will do.